[WIP] Lokiless netobserv (#72)
netobserv/documents: blogs/lokiless_netobserv/index.md

```yaml
loki:
  enable: false
```
When configured as above, Network Observability's Prometheus metrics will continue to get scraped by OpenShift's cluster Prometheus without any additonal configuration and Network Traffic console will use Prometheus as a source for fetching the data.

## Performance and Resource utilization gains

### Query performance:
Prometheus queries are blazing fast compared to Loki queries, but don't take my word for it, let's look at the data from the query performance tests: 

Test bench environment:

* Test: We did identical 50 queries for 3 separate time ranges to render topology view for both Loki and Prometheus. Such query requests all K8s Owners for the workload running in an OpenShift Cluster that had  network flows associated to them. Since we didn't have any applications running, it was all Infrastructure workloads generating network traffic. In Network Observability such unfiltered view will have topology rendered as below: 

    ![unfiltered topology view](images/owner_screenshot.png)

* Cluster config: 3 worker and 3 master nodes, AWS m5.2xlarge machines

> * LokiStack size: 1x.extra-small

Reviewer: that's probably a mistake to use extra small here:
 ```
This size is ultimately dedicated for development and demo purposes.
```

Author: I'll re-run the test with 1x.small and gather metrics, with automation shouldn't take very long :)

Reviewer: awesome ! thanks ! We'll see if it's better (or not !)

Results:

  Below table shows 90th Percentile query times for each table:

  | Time Range | Loki      | Prometheus
  | :--------: | :-------: | :----------:
  | Last 5m    | 2287 ms   | 95.5 ms
  | Last 1h    | 4581 ms   | 236 ms
  | Last 6h    | > 10 s    | 394 ms

> As time range to fetch network flows gets wider, Loki queries tends to get slower or timing out, while Prometheus queries is able render the data within fraction of a second.

Reviewer: How did you get these numbers ? Is it from your browser networking tab ?

Author: @jpinsonneau I wrote a small script that does 50 queries for each Prometheus and Loki: https://github.com/memodi/openshift-tests-private/commit/4ea4683753c1e6e945ecde0dcc0ee871270b6920 and after test is done gathered prom. metrics for loki and prom duration calls with query like:
```
histogram_quantile(0.9, sum(rate(netobserv_loki_calls_duration_bucket{code="200"}[5m])) by (le, code))
```

Reviewer: that's cool ! You should link the test in the blog :smile_cat:

1. Test bed 1: node-density-heavy workload ran against 25 nodes cluster.
2. Test bed 2: ingress-perf workload ran against 65 nodes cluster.
3. Test bed 3: cluster-densit-v2 workload ran against 120 nodes cluster

Below graphs shows total vCPU and memory usage for a recommended Network Observability stack (flowlogs-pipeline, eBPF-agent, Kafka and optionally Loki) for production clusters.

![Compare total vCPUs utilized with and without Loki](<images/Total vCPUs consumed.png>)
![Compare total RSS utilized with and without Loki](<images/Total Memory (RSS) consumed.png>)

<sup>*</sup> actual resource utilization may depend on various factors such as flowcollector sampling size, number of workloads and nodes in an OCP cluster

## Trade-offs:
We saw having Prometheus as datasource provides impressive performance gains and sub-second query times, however it introduces below constraints:

1. Without storage of network flows it no longer provides Traffic flows table. <TODO: insert a picture Traffic table greyed out>

> 2. Per-pod level of resource granularity is not available since it causes Prometheus metrics to have high cardinality. <TODO: insert a picture where diff between with-Loki and without-Loki Scope>

Reviewer: it would be interesting to also compare the output storage size with and without loki

Author: what do you mean by output storage size? I have mentioned here that users don't need to provision additional storage: 
https://github.com/netobserv/documents/pull/72/files#diff-77e07c919145d98ff5ecc61c81fe91d345f160619e8ddbf9e39d4dab956d7921R43

Reviewer: Prometheus storage increases when enabling netobserv metrics.
There are even capabilities to export  to remote storages: https://prometheus.io/docs/prometheus/latest/storage/#remote-storage-integrations

The goal here would be to be able to say how much netobserv metrics could take (depending of what you enabled of course) compared to Loki storage when storing every flows

Author: > Prometheus storage increases when enabling netobserv metrics.

Do you think it would cause significant increase in prometheus storage by having these metrics published?

Reviewer: It could depending of the cardinality of the selected metrics. But it's also important to highlight low impact configurations :smile_cat:

Author: @jotak do you have prometheus metrics to measure its storage util.? the ones on Prometheus dashboard for head_series and head_chunks - are those for prometheus in-memory storage for metrics? I did not find any pvc set for prometheus in OCP cluster.

Reviewer: I found that estimation on grafana website:
```
The Prometheus metrics resource usage depends mainly on the number of active series that need to be scraped and the scrape interval.

As a rule of thumb, per each 1 million active series and with the default scrape interval, you can expect to use approximately:

0.4 CPU cores
11 GiB of memory
1.5 MiB/s of total network bandwidth, send and receive
```

https://grafana.com/docs/agent/latest/flow/tasks/estimate-resource-usage/#prometheus-metrics

Author: @jotak @jpinsonneau I did some testing to figure prometheus impact by:
```
when netobserv is deployed:
head series/chunks - jumps from ~470K to ~1M

Storage: `prometheus_tsdb_wal_storage_size_bytes` which shows high rate of increase [1] (not sure though if this is metric is indicator of prometheus' storage
Prom. CPU: doubles from 0.15 to 0.3 [2]
Prom. RSS:  jumps from 2.5G to 4.7 [3]
```
[1]
<img width="1201" alt="prom_wal_rate" src="https://github.com/user-attachments/assets/36be0632-843e-4682-990a-a1442809a805">

[2]
<img width="1192" alt="prom_cpu_usage" src="https://github.com/user-attachments/assets/faa5a3d8-ff64-4235-8f2f-a2121f25a2b5">

[3]
<img width="1187" alt="prom_memory_usage" src="https://github.com/user-attachments/assets/30b7a143-dd8c-4be3-8d8e-00771dd8abd9">

This is when default metrics are published when loki is disabled. 
I am not sure though this is the absolute difference between with and without Loki since many of the metrics we already published even when Loki is enabled.

Reviewer: I think to see the total storage size you should use `prometheus_tsdb_storage_blocks_bytes: The number of bytes that are currently used for local storage by all blocks`.

You can also have a look to `prometheus_tsdb_head_chunks_storage_size_bytes: Size of the chunks_head directory` in parallel.

Anyways these results are not bad but we should inform the customers about these in the doc.

Reviewer 2: @memodi when you say:
> Prom. CPU: doubles from 0.15 to 0.3 [2]
> Prom. RSS:  jumps from 2.5G to 4.7 [3]

Is it before, after, or during the NDH workloads being generated?

Author: @jotak it's during the NDH runs. CPU returns to the previous levels after the workload is completed, however RSS jump is more stickier probably because of those active series are still held in prometheus memory.

@jpinsonneau for some reason `prometheus_tsdb_storage_blocks_bytes` is always 0 in default config.

Reviewer 2: @memodi so maybe it's also worth check what's the memory overhead if we run NDH without netobserv at all, because creating pods etc. does have an impact on metrics even without netobserv

   
   Should you need per-flow or per-pod level of granularity for diagnostic and troubleshooting needs, other than enabling Loki you have multiple other options available:

   a. Collect flowlogs into your preferred data analytics tool using `.spec.exporters` config in Flowcollector, currently Kafka and IPFIX are supported exporters.

   b. In this release, Network Observability also introduced `FlowMetrics` API which lets you create custom metrics which may not be available out of the box. `FlowMetrics` API creates on-demand Prometheus metrics based on enriched flowlogs fields which can be used as labels for custom Prometheus metrics. _Note: Be careful with this option though, introducing metrics that may have labels with high cardinality increases cluster's Promethes resource usage and may impact overall cluster monitoring_.

> 3. Restricted multi-tenancy - Currently Prometheus in OpenShift cluster currently doesn't support multi-tenancy in a way that Loki does, non-admin users can be added to `cluster-monitoring-view` where user will have access to view all available Prometheus metrics.

Reviewer 2: "currently" is repeated

```suggestion
3. Restricted multi-tenancy - Prometheus in OpenShift cluster currently doesn't support multi-tenancy in a way that Loki does, non-admin users can be added to `cluster-monitoring-view` where user will have access to view all available Prometheus metrics.
```

> # Network Observability without Loki

Reviewer 3: Just my opinion and an observation of how we use the same kinds of words to describe different aspects of the NetObserv toolset, but I think the "without Loki" part is still important here and also "Operator". Maybe "Light-weight Network Observability Operator without Loki". In the docs, we also describe the CLI as light-weight Network Observability, but it _the_ Operator.

By: Mehul Modi, Steven Lee

Recently, the Network Observability Operator released version  1.6, which added a major enhancement to provide network insights for your OpenShift cluster without Loki. This enhancement was also featured in [What's new in Network Observability 1.6](../whats_new_1.6) blog, providing a quick overview of the feature. In this blog, lets look at some of the advantages and trade-offs users would have when deploying the Network Observability Operator with Loki disabled. As more metrics are enabled by default with this feature, we'll also demonstrate a use-case on how those metrics can benefit users for real world scenarios.

> 

Author: ```suggestion
Recently, the Network Observability Operator released version  1.6, which added a major enhancement to provide network insights for your OpenShift cluster without Loki. This enhancement was also featured in [What's new in Network Observability 1.6](https://developers.redhat.com/articles/2024/08/12/whats-new-network-observability-16) blog, providing a quick overview of the feature. Until this release, Loki was required to be deployed alongside Network Observability to store the network flows data. In this blog, lets look at some of the advantages and trade-offs users would have when deploying the Network Observability Operator with Loki disabled. As more metrics are enabled by default with this feature, we'll also demonstrate a use-case on how those metrics can benefit users for real world scenarios.

```

how about this text here to say how Loki was used.
cc @skrthomas

> 

Reviewer 3: That's a good point. I wonder if we can link to Joel's other blog(https://github.com/netobserv/documents/blob/main/blogs/lokiless/index.md) somewhere here; I think there's a nice historical deep dive there. Maybe something like this: 
> Recently, the Network Observability Operator released version  1.6, which added a major enhancement to provide network insights for your OpenShift cluster without Loki. This work builds on an effort to reduce reliance on Loki, which began starting with 1.4 release of Network Observability.

> 

Reviewer 3: and then the link somewhere in that sentence. I'm sure that blog is published somewhere just gotta find it. I could only find the GitHub link

> 

Reviewer 3: Looks like our comments crossed in the ether, but your suggestion looks good too Mehul. My only question is whether or not we want to mention that this reduced dependency on Loki has been a WIP for a couple releases now? I know 1.6 is the most robst version of this so maybe we don't want to mention anything about it.

> 

Author: I think it's okay to leave out the history from this.

# Configure Network Observability without Loki
Loki as datasource is currently enabled by default. To configure Network Observability operator without Loki, when configuring Flowcollector resource simply set `.spec.loki.enable` to `false`

```yaml
loki:
  enable: false
```

> When configured as above, Network Observability's Prometheus metrics will continue to get scraped by OpenShift's cluster Prometheus without any additonal configuration and Network Traffic console will use Prometheus as a source for fetching the data.

Author: thanks, that's great insights.

## Performance and Resource utilization gains

> 

Author: added, can you PTAL?

> 

Reviewer 3: Lgtm, thanks Mehul :)

### Query performance:
Prometheus queries are blazing fast compared to Loki queries, but don't take my word for it, let's look at the data from the query performance tests: 

Test bench environment:

* **Test**: We conducted 50 identical queries for 3 separate time ranges to render a topology view for both Loki and Prometheus. Such a query requests all K8s Owners for the workload running in an OpenShift Cluster that had  network flows associated to them. Since we did not have any applications running, only Infrastructure workloads generated network traffic. In Network Observability such an unfiltered view renders topology rendered as follows: 

>     ![unfiltered topology view](images/owner_screenshot.png)

Author: I was kinda dubious about it for the same reason, I removed the image and also updated the text of this para.

* Test bed: 9 worker and 3 master nodes, AWS m5.2xlarge machines
* LokiStack size: 1x.small

Results:

  Below table shows 90th Percentile query times for each table:

  | Time Range | Loki      | Prometheus
  | :--------: | :-------: | :----------:
  | Last 5m    | 984 ms    | 99 ms
  | Last 1h    | 2410 ms   | 236 ms
  | Last 6h    | > 10 s    | 474 ms

> As time range to fetch network flows gets wider, Loki queries tends to get slower or timing out, while Prometheus queries is able render the data within fraction of a second.

Reviewer 3: Some verb tense items

### Resource utilization:

> In our tests conducted on 3 different test beds with varied workloads and network throughput, when Network Observability is configured without Loki, total savings of Memory usage could be in range 45-65% and CPU utilzation could be lower by 10-20%<sup>*</sup>. Not to mention you will not need to provision and plan for additional object storage in public clouds for Loki, overall reducing the cost and improving operational efficiency significantly.

Reviewer 3: ```suggestion
In our tests conducted on 3 different test beds with varied workloads and network throughput, when Network Observability is configured without Loki, total savings of memory usage are in the 45-65% range and CPU utilization is lower by 10-20%<sup>*</sup>. Not to mention you do not need to provision and plan for additional object storage in public clouds for Loki, overall reducing the cost and improving operational efficiency significantly.
```

Reviewer 3: Mostly verb tense/active voice suggestions. A question, you capitalize Memory, but I'm not sure it _needs_ to be capitalized.

> In our perf tests, [kube-burner](https://github.com/kube-burner/kube-burner) workloads were used to generate several objects and create heavy network traffic. We used sampling rate of 1 for all below tests. To further describe each test bed:

Reviewer 3: ```suggestion
In our perf tests, [kube-burner](https://github.com/kube-burner/kube-burner) workloads were used to generate several objects and create heavy network traffic. We used a sampling rate of 1 for all of the following tests:
```

Reviewer 3: suggest replacing "below" with "following" and IMO, you don't need to have the "To further describe bit". Just go ahead with the colon :)

1. Test bed 1: node-density-heavy workload ran against 25 nodes cluster.
2. Test bed 2: ingress-perf workload ran against 65 nodes cluster.

> 3. Test bed 3: cluster-density-v2 workload ran against 120 nodes cluster

Reviewer 3: Again, wondering what you think about bolding these list items **Test bed 1, 2, 3**

Author: yeah, that make sense.

> 3. Test bed 3: cluster-density-v2 workload ran against 120 nodes cluster

Author: I updated the info from the official docs here: https://docs.openshift.com/container-platform/4.12/observability/network_observability/configuring-operator.html#network-observability-total-resource-usage-table_network_observability

> Below graphs shows total vCPU, memory and storage usage for a recommended Network Observability stack  - flowlogs-pipeline, eBPF-agent, Kafka, Prometheus and optionally Loki for production clusters.

Reviewer 3: ```suggestion
The following graphs show total vCPU, memory and storage usage for a recommended Network Observability stack  - flowlogs-pipeline, eBPF-agent, Kafka, Prometheus and optionally Loki for production clusters.
```

![Compare total vCPUs utilized with and without Loki](<blogs/lokiless_netobserv/images/vCPUs consumed by NetObserv stack.png/Total vCPUs consumed.png>)

> ![Compare total RSS utilized with and without Loki](<blogs/lokiless_netobserv/images/Memory consumed by NetObserv stack.png>)

Author: updated the links

> Let's look at the amount of estimated storage you'd need for all the network flows and Prometheus metrics that Network Observability has to store. For context, even when Loki is installed Network Observability publishes default set of Prometheus metrics for monitoring dashboards, and it adds additional metrics when Loki is disabled to visualize network flows. The graphs below shows the estimated amount of storage required to store 15 days of Network flows (when configured with Loki), Prometheus metrics and Kafka as intermediary data streaming layer between eBPF-agent and flowlogs-pipeline. 

Reviewer 3: ```suggestion
Let's look at the amount of estimated storage you need for all the network flows and Prometheus metrics that Network Observability has to store. For context, even when Loki is installed, Network Observability publishes a default set of Prometheus metrics for monitoring dashboards, and it adds additional metrics when Loki is disabled to visualize network flows. The graphs below show the estimated amount of storage required to store 15 days of network flows data (when configured with Loki), with Prometheus metrics and Kafka as an intermediary data streaming layer between the eBPF-agent and the flowlogs-pipeline. 
```

Reviewer 3: Just verb tense and adding some the's and a's. One question, "network flows" only or "network flows data"?

Author: > "network flows" only or "network flows data"?
 
either should be fine.  Although I am not sure about "with Prometheus" here, wdyt?

Reviewer 3: I think the "with" or alternatively "in addition to" is needed to introduce it because its a non-restrictive clause that is independent of the main sentence, but still a fragment.

The network flows rate for each test bed was 10K, 13K, 30K flows/second respectively. The storage for Loki includes AWS S3 bucket utilization and its PVC usage. For Kafka PVC storage value, it assumes 1 day of retention or 100 GB whichever is attained first.

![Compare total Storage utilized with and without Loki](<blogs/lokiless_netobserv/images/15 days Storage consumption.png>)

> As seen across test beds above, we find storage savings of 90% when Network Observability is configured without Loki.

Reviewer 3: ```suggestion
As seen across the test beds, we find a storage savings of 90% when Network Observability is configured without Loki.
```

> <sup>*</sup> actual resource utilization may depend on various factors such as network traffic, flowcollector sampling size, number of workloads and nodes in an OCP cluster

Reviewer 3: ```suggestion
<sup>*</sup> actual resource utilization may depend on various factors such as network traffic, FlowCollector sampling size, and the number of workloads and nodes in an OpenShift Container Platform cluster
```

> <sup>*</sup> actual resource utilization may depend on various factors such as network traffic, FlowCollector sampling size, and the number of workloads and nodes in an OpenShift Container Platform cluster

Author: I have mentioned it here about CPU and Memory savings in intro paragraph: https://github.com/netobserv/documents/pull/72/files#diff-77e07c919145d98ff5ecc61c81fe91d345f160619e8ddbf9e39d4dab956d7921R42 , can you add suggestions what else we could say here?

## Trade-offs:

> We saw having Prometheus as datasource provides impressive performance gains and sub-second query times, however it introduces below constraints:

Reviewer 3: ```suggestion
We saw having Prometheus as datasource provides impressive performance gains and sub-second query times, however it introduces the following constraints:
```

> 1. Without storage of network flows it no longer provides Traffic flows table.

Reviewer 3: ```suggestion
1. Without storage of network flows data, the Network Observability OpenShift web console no longer provides the Traffic flows table.
```

![Disabled table view](images/disabled_table_view.png)

2. Per-pod level of resource granularity is not available since it causes Prometheus metrics to have high cardinality.

![Topology scope changes](images/topology_scope.png)
   

>    Should you need per-flow or per-pod level of granularity for diagnostic and troubleshooting needs, other than enabling Loki you have multiple other options available:

Reviewer 3: ```suggestion
   Should you need per-flow or per-pod level of granularity for diagnostic and troubleshooting, other than enabling Loki you have multiple other options available:
```

Reviewer 3: Since you already say, "Should you need", this needs can be removed.

>    a. Collect flowlogs into your preferred data analytics tool using `.spec.exporters` config in Flowcollector, currently Kafka and IPFIX are supported exporters.

Reviewer 3: ```suggestion
   a. Collect flowlogs into your preferred data analytics tool using the`.spec.exporters` configuration in the FlowCollector. Currently Kafka and IPFIX are supported exporters.
```

>    b. In this release, Network Observability also introduced `FlowMetrics` API which lets you create custom metrics which may not be available out of the box. `FlowMetrics` API creates on-demand Prometheus metrics based on enriched flowlogs fields which can be used as labels for custom Prometheus metrics. _Note: Be careful with this option though, introducing metrics that may have labels with high cardinality increases cluster's Promethes resource usage and may impact overall cluster monitoring_.

Reviewer 3: ```suggestion
   b. In this release, Network Observability also introduced the `FlowMetrics` API, which lets you create custom metrics that are not available out of the box. The `FlowMetrics` API creates on-demand Prometheus metrics based on enriched flowlogs fields, which can be used as labels for custom Prometheus metrics. _Note: Use this option with caution, as introducing metrics that may have labels with high cardinality increases the cluster's Prometheus resource usage and might impact overall cluster monitoring_.
```

Reviewer 3: - custom metrics that are not available out of the box. I guess, I wasn't aware of a "maybe" in this feature.
- Suggesting alternative phrasing for the warning.

> 3. Restricted multi-tenancy - Prometheus in OpenShift cluster currently doesn't support multi-tenancy in a way that Loki does, non-admin users can be added to `cluster-monitoring-view` where user will have access to view all available Prometheus metrics.

Reviewer 3: ```suggestion
3. Restricted multi-tenancy - Prometheus in OpenShift cluster currently doesn't support multi-tenancy in a way that Loki does. Non-admin users can be added to `cluster-monitoring-view` where the user can have access to view all available Prometheus metrics.
```

Let's look at a scenario how users can benefit from metrics published by Network Observability Operator. For instance, if you suspect anomaly with DNS lookups in your cluster and want to investigate workloads that may be facing DNS latencies. With Network Observability's `DNSTracking` feature and enriched Prometheus metrics you can quickly set up an alert to trigger on high DNS latencies.

For example, below alert will trigger for any workloads that experiences DNS latency > 100ms: 
```yaml
apiVersion: monitoring.coreos.com/v1
kind: PrometheusRule
metadata:
  name: dns-latency-alert
  namespace: netobserv
spec:
  groups:
  - name: DNSLatencyAlert
    rules:
    - alert: DNSLatencyAlert
      annotations:
        message: |-
          {{ $labels.DstK8S_OwnerName }} in {{ $labels.DstK8S_Namespace }} is experiencing high DNS Latencies.
        summary: "Trigger for any workloads experiencing > than 100ms DNS Latency."
      expr: histogram_quantile(0.9, sum(rate(netobserv_workload_dns_latency_seconds_bucket{DstK8S_Namespace!=""}[2m])) by (le,DstK8S_Namespace,DstK8S_OwnerName))*1000 > 100
      for: 10s
      labels:
        severity: warning
```

> To demonstrate this use-case, I configured CoreDNS's [erratic plugin](https://coredns.io/plugins/erratic/) in openshift-dns to add latencies for `example.org` domain using below config:

Reviewer 3: is `openshift-dns` a namespace? if it is, I suggest backticks and calling it a namespace.

```
example.org {
        erratic {
            delay 2 100ms
        }
}
```

> Above config add 100ms delay to every 2nd DNS request coming in for example.org. A test pod performing DNS lookups for `example.org` every 1 second was created, eventually triggering earlier configured `DNSLatencyAlert` in my OCP cluster.

Reviewer 3: Instead of saying "above" its clearer to say what the configuration is.

![DNSLatency alert triggered for threshold > 100ms](images/dns_latency_alert_firing.png)

Similarly, additional alerts on different DNS response codes could be set up, for example an alert for DNS lookup failures such as DNS queries receiving NXDOMAIN or SERVFAIL responses can also be set up as flowlogs and metrics are already enriched with DNS response codes.

In addition to metrics for `DNSTracking` feature, Network Observability provides metrics for other features such as Round-Trip-Times and Packet Drops as well.

## Conclusion and next steps:

> Network Observability operator provides the visibility you need to proactively detect issues within OpenShift cluster networking. Now with an option to disable loki, Network Observability operator provides light weight solution to visualize, diagnose and troubleshoot networking issues faster at a lower cost. Network Observability's Prometheus metrics can be leveraged to set up user defined alerts in your OCP cluster.

Author: > to set up user defined alerts in your OCP cluster

about this text, does this sound like a compulsion that user has to create alerts? I want to make it sound like on the lines of "and instead leverage Prometheus metrics which can be optionally used to set up user defined alerts in your OCP cluster", wdyt? or the sentence is getting too complex if we add that part?

> Network Observability operator provides the visibility you need to proactively detect issues within OpenShift cluster networking. Now with an option to disable loki, Network Observability operator provides light weight solution to visualize, diagnose and troubleshoot networking issues faster at a lower cost. Network Observability's Prometheus metrics can be leveraged to set up user defined alerts in your OCP cluster.

Reviewer 3: You're right. The "to set up" is definitely seems more like the customer needs to do it. I wonder if the wording more along the lines of "the ability" to do these things would be more appropriate? I hesitate to use "option" again since we just use it in the "Now with an option to disable Loki". Maybe something like this:

> Now with an option to disable Loki, and the ability to leverage Prometheus metrics to set up user defined alerts in your OCP cluster, the Network Observability Operator provides a light-weight solution to visualize, diagnose and troubleshoot networking issues faster and at a lower cost.

> Network Observability operator provides the visibility you need to proactively detect issues within OpenShift cluster networking. Now with an option to disable loki, Network Observability operator provides light weight solution to visualize, diagnose and troubleshoot networking issues faster at a lower cost. Network Observability's Prometheus metrics can be leveraged to set up user defined alerts in your OCP cluster.

Author: how about we just leave out the alerts part in the conclusion, it's one of the use cases of how users can leverage these metrics, but not a something we're providing:
```suggestion
The Network Observability Operator provides the visibility you need to proactively detect issues within OpenShift cluster networking. Now with an option to disable Loki, and instead leverage Prometheus metrics, the Network Observability Operator provides a light-weight solution to visualize, diagnose and troubleshoot networking issues faster and at a lower cost. 
```

# Network Observability without Loki

By: Mehul Modi, Steven Lee

Recently, Network Observability operator 1.6 released a major enhancement to provide network insights for your OpenShift cluster without Loki. This enhancement was also featured in [What's new in Network Observability 1.6](../whats_new_1.6) blog providing quick overview of the feature. In this blog, lets look at some of the advantages and trade-offs users would have when deploying network observability with Loki disabled.

# Configure Network Observability without Loki
Loki as datasource is currently enabled by default. To configure Network Observability operator without Loki, when configuring Flowcollector resource simply set `.spec.loki.enable` to `false`

```yaml
loki:
  enable: false

> ```

Reviewer 2: You could also add some context, such as telling that, without Loki, metrics are still being sent to Prometheus - and since Prometheus is in core OpenShift payload, you don't need further backend configuration.

In our perf tests, [kube-burner](https://github.com/kube-burner/kube-burner) workloads were used to generate several objects and create heavy network traffic. We used sampling rate of 1 for all below tests. To further describe each test bed:

1. Test bed 1: node-density-heavy workload ran against 25 nodes cluster.
2. Test bed 2: ingress-perf workload ran against 65 nodes cluster.
3. Test bed 3: cluster-densit-v2 workload ran against 120 nodes cluster

Below graphs shows total vCPU and memory usage for a recommended Network Observability stack (flowlogs-pipeline, eBPF-agent, Kafka and optionally Loki) for production clusters.

![Compare total vCPUs utilized with and without Loki](<images/Total vCPUs consumed.png>)
![Compare total RSS utilized with and without Loki](<images/Total Memory (RSS) consumed.png>)

<sup>*</sup> actual resource utilization may depend on various factors such as flowcollector sampling size, number of workloads and nodes in an OCP cluster

> This comes with couple of trade-offs though, without storage of network flows it no longer provides Traffic flows table. Also, per-pod level of resource granularity is not available since it causes prometheus metrics to have high cardinality. However, should you need per-pod or per-flow level of granularity for diagnostic and troubleshooting needs, enabling loki should be pretty straightforward and both datasources can be used. When both loki and prometheus datasources are enabled, while querying on Netflow Traffic page, prometheus queried data over will be used wherever possible since it offers faster performance.

Author: Added exporters and FlowMetrics API as potential options alongside Tradeoffs

> 

Reviewer 2: Maybe there should be a whole section about the tradoff, that includes the previous paragraph about the level of details + the multi-tenancy aspects.

Also speaking of trade-offs, I don't know if you want to dive in these details, but there's also the fact that the metrics model imposes limitations in how we can combine filters/aggregations. For instance, because the per-namespace bandwidth and the per-node bandwidth are provided by two separate metrics, we cannot query with namespace filters and Node scope, or with node filter and Namespace scope. Maybe that's something to mention, which still gives an advantage to Loki in terms of querying flexibility.

Author: I have added a separate section for Trade-offs. I think we can leave out filtering/aggregation limitation details, because it's more of a implementation detail which we eventually plan to improve UX for.

## Accessing Netflow Traffic for non-admins:

> While prometheus currently doesn't supports multi-tenancy in a way that Loki does in an OpenShift cluster, non-admin users can be added to `cluster-monitoring-view`. For example, below command can be used to enable prometheus metrics visualizing for `testuser-0` user.

Reviewer 2: ```suggestion
While Prometheus currently doesn't support multi-tenancy in a way that Loki does in an OpenShift cluster, non-admin users can be added to `cluster-monitoring-view`. For example, below command can be used to enable prometheus metrics visualizing for `testuser-0` user.
```

`oc adm policy add-cluster-role-to-user cluster-monitoring-view  testuser-0`

## Network Observability metrics use case:

> Let's look at a scenario how users can benefit from metrics published by Network Observability Operator. For instance, if you suspect anamaly with DNS lookups in your cluster and want to investigate workloads that may be facing DNS latencies. With Network Observability's `DNSTracking` feature and enriched prometheus metrics you can quickly set up an alert to trigger on high DNS latencies.

Reviewer 2: ```suggestion
Let's look at a scenario how users can benefit from metrics published by Network Observability Operator. For instance, if you suspect anomaly with DNS lookups in your cluster and want to investigate workloads that may be facing DNS latencies. With Network Observability's `DNSTracking` feature and enriched prometheus metrics you can quickly set up an alert to trigger on high DNS latencies.
```

For example, below alert will trigger for any workloads that experiences DNS latency > 100ms: 
```yaml
apiVersion: monitoring.coreos.com/v1
kind: PrometheusRule
metadata:

>   name: dns-latency-alert

Author: I don't think it needs to be in particular NS, but I just added this `netobserv` NS since to just organize resource better. I'll check if it's able to trigger when deployed in different NS as well.

spec:
  groups:
  - name: DNSLatencyAlert
    rules:
    - alert: DNSLatencyAlert
      annotations:
        message: |-

>           {{ $labels.DstK8S_OwnerName }} in {{ $labels.DstK8S_Namespace }} is experiencing DNS Latencies.

Reviewer 2: ```suggestion
          {{ $labels.DstK8S_OwnerName }} in {{ $labels.DstK8S_Namespace }} is experiencing high DNS Latencies.
```
?

        summary: "Trigger for any workloads experiencing > than 100ms DNS Latency."
      expr: topk(7, (histogram_quantile(0.9, sum(rate(netobserv_workload_dns_latency_seconds_bucket{SrcK8S_Namespace!=""}[2m])) by (le,SrcK8S_Namespace,SrcK8S_OwnerName,DstK8S_Namespace,DstK8S_OwnerName))*1000> 100) or (histogram_quantile(0.9, sum(rate(netobserv_workload_dns_latency_seconds_bucket{DstK8S_Namespace!=""}[2m])) by (le,SrcK8S_Namespace,SrcK8S_OwnerName,DstK8S_Namespace,DstK8S_OwnerName))*1000 > 100))

>       for: 10s

Reviewer 2: Not 100% sure but I think here you can keep only the second part of the query, to simplify? DNS latencies are on responses and we're displaying "{{ $labels.DstK8S_OwnerName }} in {{ $labels.DstK8S_Namespace }}" so I think we can just look at "{DstK8S_Namespace!=""}"

Also I'd remove the topk for an alert query.

```
histogram_quantile(0.9, sum(rate(netobserv_workload_dns_latency_seconds_bucket{DstK8S_Namespace!=""}[2m])) by (le,SrcK8S_Namespace,SrcK8S_OwnerName,DstK8S_Namespace,DstK8S_OwnerName))*1000 > 100
```

![DNSLatency alert triggered for threshold > 100ms](images/dns_latency_alert_firing.png)

Similarly, additional alerts on different DNS response codes could be set up, for example an alert for DNS lookup failures such as DNS queries receiving NXDOMAIN or SERVFAIL can also be set up as flowlogs and metrics are already enriched with DNS response codes.

In addition to metrics for `DNSTracking` feature, Network Observability provides metrics for other features such as Round-Trip-Times and Packet Drops as well.

## Conclusion and next steps:

Network Observability operator provides the visibility you need to proactively detect issues with OpenShift cluster networking. Now with an option to disable loki, Network Observability operator provides light weight solution to visualize, diagnose and troubleshoot networking issues faster at a lower cost. Network Observability's prometheus metrics can be leveraged to set up user defined alerts in your OCP cluster.

> While some feature parity gaps exists when compared to configuration with Loki enabled, team is actively working to close that gap and enhance this feature by enabling visualization for packet drops and better prometheus muti-tenancy.

Reviewer 2: ```suggestion
While some feature parity gap exists when compared to configuration with Loki enabled, team is actively working to close that gap and enhance this feature by enabling visualization for packet drops and better Prometheus multi-tenancy.
```
